Source of the population data: https://seer.cancer.gov/popdata/singleages.html

In [14]:
import pandas as pd

pd.set_option("mode.copy_on_write", True)

In [1]:
# Define the column widths for each field in the fixed-length data
column_widths = [4, 2, 2, 3, 2, 1, 1, 1, 2, 8]
# Define the column names
column_names = ['Year', 'State', 'State_FIPS', 'County_FIPS', 'Registry', 'Race', 'Origin', 'Sex', 'Age', 'Population']

# Read the fixed-length data file with column names
data = pd.read_fwf('us.1969_2020.19ages.adjusted.txt', widths=column_widths, names=column_names)

# Display the data
print(data)

          Year State  State_FIPS  County_FIPS  Registry  Race  Origin  Sex  \
0         1969    AL           1            1        99     1       9    1   
1         1969    AL           1            1        99     1       9    1   
2         1969    AL           1            1        99     1       9    1   
3         1969    AL           1            1        99     1       9    1   
4         1969    AL           1            1        99     1       9    1   
...        ...   ...         ...          ...       ...   ...     ...  ...   
15536023  2020    WY          56           45        99     3       9    2   
15536024  2020    WY          56           45        99     3       9    2   
15536025  2020    WY          56           45        99     3       9    2   
15536026  2020    WY          56           45        99     3       9    2   
15536027  2020    WY          56           45        99     3       9    2   

          Age  Population  
0           0         159  
1      

In [2]:
# Tried to use dask to read the data, but it did not get much faster
# import dask.dataframe as dd
# import pandas as pd

# # Define the column widths for each field in the fixed-length data
# column_widths = [4, 2, 2, 3, 2, 1, 1, 1, 2, 8]
# # Define the column names
# column_names = ['Year', 'State', 'State_FIPS', 'County_FIPS', 'Registry', 'Race', 'Origin', 'Sex', 'Age', 'Population']

# ddf = dd.read_fwf('us.1969_2020.19ages.adjusted.txt', widths=column_widths, names=column_names)
# data = ddf.compute()

In [3]:
data

Year State  State_FIPS  County_FIPS  Registry  Race  Origin  Sex  \
0         1969    AL           1            1        99     1       9    1   
1         1969    AL           1            1        99     1       9    1   
2         1969    AL           1            1        99     1       9    1   
3         1969    AL           1            1        99     1       9    1   
4         1969    AL           1            1        99     1       9    1   
...        ...   ...         ...          ...       ...   ...     ...  ...   
15536023  2020    WY          56           45        99     3       9    2   
15536024  2020    WY          56           45        99     3       9    2   
15536025  2020    WY          56           45        99     3       9    2   
15536026  2020    WY          56           45        99     3       9    2   
15536027  2020    WY          56           45        99     3       9    2   

          Age  Population  
0           0         159  
1           1         657  
2           2        1137  
3           3         956  
4           4         721  
...       ...         ...  
15536023   14          21  
15536024   15           2  
15536025   16           2  
15536026   17           2  
15536027   18           2  

[15536028 rows x 10 columns]

In [8]:
count_combinations = data.groupby(['State_FIPS', 'County_FIPS']).size().reset_index(name='Count')
print(count_combinations)

      State_FIPS  County_FIPS  Count
0              1            1   5563
1              1            3   5900
2              1            5   5498
3              1            7   4969
4              1            9   5628
...          ...          ...    ...
3160          56           39   4169
3161          56           41   4448
3162          56           43   4373
3163          56           45   3875
3164          99          999    114

[3165 rows x 3 columns]


In [11]:
year_count = data.groupby(['Year']).size().reset_index(name='Count')
year_count

Year   Count
0   1969  258492
1   1970  262596
2   1971  265513
3   1972  270368
4   1973  273392
5   1974  275029
6   1975  275959
7   1976  278207
8   1977  280137
9   1978  281855
10  1979  280156
11  1980  281356
12  1981  284337
13  1982  287421
14  1983  292700
15  1984  292668
16  1985  296620
17  1986  296636
18  1987  296637
19  1988  294018
20  1989  292506
21  1990  278085
22  1991  278136
23  1992  278212
24  1993  278259
25  1994  278645
26  1995  278647
27  1996  278753
28  1997  278759
29  1998  278842
30  1999  278938
31  2000  311492
32  2001  311721
33  2002  313340
34  2003  315721
35  2004  317109
36  2005  319800
37  2006  321346
38  2007  322143
39  2008  321634
40  2009  321778
41  2010  325506
42  2011  327073
43  2012  328804
44  2013  330322
45  2014  331671
46  2015  332959
47  2016  334073
48  2017  335125
49  2018  336132
50  2019  337508
51  2020  338892

In [16]:
selected_years = data[(data['Year'] >= 2002) & (data['Year'] <= 2016)]
selected_years


Year State  State_FIPS  County_FIPS  Registry  Race  Origin  Sex  \
9325092   2002    AL           1            1        99     1       9    1   
9325093   2002    AL           1            1        99     1       9    1   
9325094   2002    AL           1            1        99     1       9    1   
9325095   2002    AL           1            1        99     1       9    1   
9325096   2002    AL           1            1        99     1       9    1   
...        ...   ...         ...          ...       ...   ...     ...  ...   
14188366  2016    WY          56           45        99     3       9    2   
14188367  2016    WY          56           45        99     3       9    2   
14188368  2016    WY          56           45        99     3       9    2   
14188369  2016    WY          56           45        99     3       9    2   
14188370  2016    WY          56           45        99     3       9    2   

          Age  Population  
9325092     0         211  
9325093     1         961  
9325094     2        1469  
9325095     3        1689  
9325096     4        1370  
...       ...         ...  
14188366   13          26  
14188367   14           3  
14188368   15           3  
14188369   16           2  
14188370   17           2  

[4863279 rows x 10 columns]

In [17]:
selected_years['FIPS'] = selected_years['State_FIPS'].astype(str).str.zfill(2) + selected_years['County_FIPS'].astype(str).str.zfill(3)

In [18]:
collapsed_data = selected_years.groupby(['FIPS', 'Year', 'State', 'State_FIPS', 'County_FIPS']).sum().reset_index()
collapsed_data = collapsed_data[['FIPS', 'Year', 'State', 'State_FIPS', 'County_FIPS', 'Population']]

collapsed_data

FIPS  Year State  State_FIPS  County_FIPS  Population
0      01001  2002    AL           1            1       45909
1      01001  2003    AL           1            1       46800
2      01001  2004    AL           1            1       48366
3      01001  2005    AL           1            1       49676
4      01001  2006    AL           1            1       51328
...      ...   ...   ...         ...          ...         ...
47102  56045  2013    WY          56           45        7132
47103  56045  2014    WY          56           45        7134
47104  56045  2015    WY          56           45        7202
47105  56045  2016    WY          56           45        7228
47106  99999  2005    KR          99          999      203937

[47107 rows x 6 columns]

In [20]:
melted_data = collapsed_data.pivot_table(index=['FIPS', 'State', 'State_FIPS', 'County_FIPS'], columns='Year', values='Population').reset_index()
melted_data.columns.name = None

melted_data

FIPS State  State_FIPS  County_FIPS      2002      2003      2004  \
0     01001    AL           1            1   45909.0   46800.0   48366.0   
1     01003    AL           1            3  147957.0  151509.0  156266.0   
2     01005    AL           1            5   28653.0   28594.0   28287.0   
3     01007    AL           1            7   21199.0   21399.0   21721.0   
4     01009    AL           1            9   52551.0   53457.0   54124.0   
...     ...   ...         ...          ...       ...       ...       ...   
3141  56039    WY          56           39   18837.0   19066.0   19467.0   
3142  56041    WY          56           41   19587.0   19480.0   19470.0   
3143  56043    WY          56           43    7988.0    7976.0    7960.0   
3144  56045    WY          56           45    6578.0    6610.0    6646.0   
3145  99999    KR          99          999       NaN       NaN       NaN   

          2005      2006      2007      2008      2009      2010      2011  \
0      49676.0   51328.0   52405.0   53277.0   54135.0   54761.0   55229.0   
1     160970.0  168121.0  172404.0  175827.0  179406.0  183121.0  186579.0   
2      28027.0   27861.0   27757.0   27808.0   27657.0   27325.0   27344.0   
3      22042.0   22099.0   22438.0   22705.0   22941.0   22858.0   22736.0   
4      54624.0   55485.0   56240.0   57055.0   57341.0   57372.0   57561.0   
...        ...       ...       ...       ...       ...       ...       ...   
3141   19632.0   20014.0   20472.0   20988.0   21232.0   21298.0   21422.0   
3142   19494.0   19709.0   20171.0   20613.0   21054.0   21090.0   20901.0   
3143    8022.0    7979.0    8169.0    8229.0    8423.0    8531.0    8451.0   
3144    6594.0    6717.0    7033.0    7133.0    7266.0    7198.0    7142.0   
3145  203937.0       NaN       NaN       NaN       NaN       NaN       NaN   

          2012      2013      2014      2015      2016  
0      54970.0   54747.0   54922.0   54903.0   55302.0  
1     190203.0  194978.0  199306.0  203101.0  207787.0  
2      27172.0   26946.0   26768.0   26300.0   25828.0  
3      22657.0   22510.0   22541.0   22553.0   22590.0  
4      57585.0   57630.0   57536.0   57535.0   57487.0  
...        ...       ...       ...       ...       ...  
3141   21643.0   22335.0   22801.0   23083.0   23255.0  
3142   21008.0   20969.0   20835.0   20777.0   20711.0  
3143    8410.0    8417.0    8277.0    8282.0    8180.0  
3144    7075.0    7132.0    7134.0    7202.0    7228.0  
3145       NaN       NaN       NaN       NaN       NaN  

[3146 rows x 19 columns]

In [23]:
rows_with_nan = melted_data[melted_data.isnull().any(axis=1)]
print(rows_with_nan['FIPS'])


76      02105
86      02195
87      02198
88      02201
90      02230
91      02232
95      02275
96      02280
3145    99999
Name: FIPS, dtype: object


* There are 9 counties with missing values during the year ranging 2002-2016, so we drop these counties
02105 02195 02198 02201 02230 02232 02275 02280 99999

In [26]:
melted_data.dropna(inplace=True)
melted_data


FIPS State  State_FIPS  County_FIPS      2002      2003      2004  \
0     01001    AL           1            1   45909.0   46800.0   48366.0   
1     01003    AL           1            3  147957.0  151509.0  156266.0   
2     01005    AL           1            5   28653.0   28594.0   28287.0   
3     01007    AL           1            7   21199.0   21399.0   21721.0   
4     01009    AL           1            9   52551.0   53457.0   54124.0   
...     ...   ...         ...          ...       ...       ...       ...   
3140  56037    WY          56           37   37428.0   37450.0   38026.0   
3141  56039    WY          56           39   18837.0   19066.0   19467.0   
3142  56041    WY          56           41   19587.0   19480.0   19470.0   
3143  56043    WY          56           43    7988.0    7976.0    7960.0   
3144  56045    WY          56           45    6578.0    6610.0    6646.0   

          2005      2006      2007      2008      2009      2010      2011  \
0      49676.0   51328.0   52405.0   53277.0   54135.0   54761.0   55229.0   
1     160970.0  168121.0  172404.0  175827.0  179406.0  183121.0  186579.0   
2      28027.0   27861.0   27757.0   27808.0   27657.0   27325.0   27344.0   
3      22042.0   22099.0   22438.0   22705.0   22941.0   22858.0   22736.0   
4      54624.0   55485.0   56240.0   57055.0   57341.0   57372.0   57561.0   
...        ...       ...       ...       ...       ...       ...       ...   
3140   38739.0   39749.0   41470.0   42358.0   44133.0   43580.0   44000.0   
3141   19632.0   20014.0   20472.0   20988.0   21232.0   21298.0   21422.0   
3142   19494.0   19709.0   20171.0   20613.0   21054.0   21090.0   20901.0   
3143    8022.0    7979.0    8169.0    8229.0    8423.0    8531.0    8451.0   
3144    6594.0    6717.0    7033.0    7133.0    7266.0    7198.0    7142.0   

          2012      2013      2014      2015      2016  
0      54970.0   54747.0   54922.0   54903.0   55302.0  
1     190203.0  194978.0  199306.0  203101.0  207787.0  
2      27172.0   26946.0   26768.0   26300.0   25828.0  
3      22657.0   22510.0   22541.0   22553.0   22590.0  
4      57585.0   57630.0   57536.0   57535.0   57487.0  
...        ...       ...       ...       ...       ...  
3140   45032.0   45189.0   44996.0   44780.0   44319.0  
3141   21643.0   22335.0   22801.0   23083.0   23255.0  
3142   21008.0   20969.0   20835.0   20777.0   20711.0  
3143    8410.0    8417.0    8277.0    8282.0    8180.0  
3144    7075.0    7132.0    7134.0    7202.0    7228.0  

[3137 rows x 19 columns]

* There are 3143 counties in the US now, and the data has 3137 rows, covered almost all the counties

In [27]:
melted_data.to_csv('population.csv', index=False)